airgun/browser.py

```diff
@@ -49,59 +65,316 @@ class BrowserPlugin(DefaultPlugin):
         '''
 
 
-def browser(browser_name=None, webdriver_name=None):
-    """Creates a webdriver browser.
+class SeleniumBrowserFactory(object):
+    """
 
-    :param browser_name: one of selenium, saucelabs, docker
-    :param webdriver_name: one of firefox, chrome, edge, ie, phantomjs
+    Usage:
+        # init factory
+        factory = SeleniumBrowserFactory(test_name=test_name)
+        # get factory browser
+        selenium_browser = factory.get_browser()
+        # navigate to desired url
+        # ...
+        # perform post-init hacks (e.g. skipping certificate error screen)
+        factory.post_init()
+        # perform your test steps
+        # ...
+        # perform factory clean-up
+        factory.finalize(passed)
 
-    If any of the params are None then will be read from properties file.
     """
 
-    webdriver_name = (webdriver_name or settings.selenium.webdriver).lower()
-    browser_name = (browser_name or settings.selenium.browser).lower()
-
-    if browser_name == 'selenium':
-        if webdriver_name == 'firefox':
-            return webdriver.Firefox()
-        elif webdriver_name == 'chrome':
-            return (
-                webdriver.Chrome() if (
-                    settings.selenium.webdriver_binary is None)
-                else webdriver.Chrome(
-                    executable_path=settings.selenium.webdriver_binary)
+    def __init__(self, browser_name=None, webdriver_name=None, test_name=None):
+        self.browser = browser_name or settings.selenium.browser or 'selenium'
```

Comment:
I would rename all these things (even if we go against conventions):
- `browser` should refer to `FF`, `IE`, `chrome`
- probably word `interface` can be used for `selenium`, `saucelabs`, `docker` and etc

Didn't dig into details, but can't we decide `browser_name or settings.selenium.browser or 'selenium'` or `webdriver_name or settings.selenium.webdriver or 'chrome'` results before these assignments. Don't we have logic for default setting values? Is it time consuming effort to implement that task?